src/library/scala/collection/Iterator.scala

```diff
@@ -416,7 +417,8 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
 
   def collect[B](pf: PartialFunction[A, B]): Iterator[B] = new AbstractIterator[B] {
     // Manually buffer to avoid extra layer of wrapping with buffered
-    private[this] var hd: A = _
+    private[this] var hd: B = _
+    private[this] final val marker = ScalaRunTime.pfMarker // this should be static if there was a way to make it
```

Comment:
Does this need to be a field? Why not a local variable in `hasNext`?

Reply:
A local variable would have to be initialized on every call to `hasNext`. This requires reading an object's field which is slower than a regular instance field. Ideally it should be a static field.

Reply:
I would suggest defining `pfMarker` in `scala.runtime.Statics` and removing this field.